File: src/main/java/net/kingdommc/darkages/numinoustreasury/command/node/session/NodeSessionProfessionRemoveCommand.java
package net.kingdommc.darkages.numinoustreasury.command.node.session;

import com.rpkit.core.service.Services;
import net.kingdommc.darkages.numinoustreasury.node.NuminousNodeCreationSession;
import net.kingdommc.darkages.numinoustreasury.node.NuminousNodeService;
import net.kingdommc.darkages.numinoustreasury.profession.NuminousProfession;
import net.kingdommc.darkages.numinoustreasury.profession.NuminousProfessionService;
import net.md_5.bungee.api.chat.ClickEvent;
import net.md_5.bungee.api.chat.ComponentBuilder;
import net.md_5.bungee.api.chat.HoverEvent;
import net.md_5.bungee.api.chat.hover.content.Text;
import org.bukkit.command.Command;
import org.bukkit.command.CommandExecutor;
import org.bukkit.command.CommandSender;
import org.bukkit.command.TabCompleter;
import org.bukkit.entity.Player;

import java.util.List;
import java.util.stream.Stream;

import static net.md_5.bungee.api.ChatColor.*;
import static net.md_5.bungee.api.chat.ClickEvent.Action.RUN_COMMAND;
import static net.md_5.bungee.api.chat.HoverEvent.Action.SHOW_TEXT;

public final class NodeSessionProfessionRemoveCommand implements CommandExecutor, TabCompleter {

    @Override
    public boolean onCommand(CommandSender sender, Command command, String label, String[] args) {
        if (!sender.hasPermission("numinoustreasury.command.node.create")) {
            sender.sendMessage(RED + "You do not have permission to create nodes.");
            return true;
        }
        if (!(sender instanceof Player player)) {
            sender.sendMessage(RED + "You must be a player to create nodes.");
            return true;
        }
        NuminousNodeService nodeService = Services.INSTANCE.get(NuminousNodeService.class);
        NuminousNodeCreationSession session = nodeService.getNodeCreationSession(player);
        if (session == null) {
            session = new NuminousNodeCreationSession();
            nodeService.setNodeCreationSession(player, session);
        }
        NuminousProfessionService professionService = Services.INSTANCE.get(NuminousProfessionService.class);
        if (args.length > 0) {
            String professionName = String.join(" ", args);
            NuminousProfession profession = professionService.getProfessionById(professionName);
            if (profession == null) {
                profession = professionService.getProfessionByName(professionName);
            }
            if (profession == null) {
                sender.sendMessage(RED + "There is no profession by that name.");
                return true;
            }
            session.removeRequiredProfessionLevel(profession);
            session.display(player);
        } else {
            player.sendMessage(WHITE + "Select a profession: ");
            professionService.getProfessions().forEach(profession -> {
                player.spigot().sendMessage(
                        new ComponentBuilder(profession.getName())
                                .color(GRAY)
                                .event(new ClickEvent(
                                        RUN_COMMAND,
                                        "/node session profession remove " + profession.getId()
                                ))
                                .event(new HoverEvent(
                                        SHOW_TEXT,
                                        new Text(
                                                new ComponentBuilder("Click here to remove " + profession.getName())
                                                        .color(GRAY)
                                                        .create()
                                        )
                                ))
                                .create()
                );
            });
        }
        return true;
    }

    @Override
    public List<String> onTabComplete(CommandSender sender, Command command, String label, String[] args) {
        NuminousProfessionService professionService = Services.INSTANCE.get(NuminousProfessionService.class);
        if (args.length == 0) {
            return professionService.getProfessions().stream()
                    .flatMap(profession -> Stream.of(
                            profession.getId(),
                            profession.getName()))
                    .toList();
        } else {
            return professionService.getProfessions().stream()
                    .flatMap(profession -> Stream.of(
                            profession.getId(),
                            profession.getName())
                    )
                    .filter(name -> name.toLowerCase().startsWith(String.join(" ", args).toLowerCase()))
                    .toList();
        }
    }
}

File: src/main/kotlin/net/arvandor/numinoustreasury/command/node/session/NodeSessionProfessionRemoveCommand.kt
package net.arvandor.numinoustreasury.command.node.session

import com.rpkit.core.service.Services
import net.arvandor.numinoustreasury.node.NuminousNodeCreationSession
import net.arvandor.numinoustreasury.node.NuminousNodeService
import net.arvandor.numinoustreasury.profession.NuminousProfessionService
import net.md_5.bungee.api.ChatColor
import net.md_5.bungee.api.chat.ClickEvent
import net.md_5.bungee.api.chat.ClickEvent.Action.RUN_COMMAND
import net.md_5.bungee.api.chat.ComponentBuilder
import net.md_5.bungee.api.chat.HoverEvent
import net.md_5.bungee.api.chat.HoverEvent.Action.SHOW_TEXT
import net.md_5.bungee.api.chat.hover.content.Text
import org.bukkit.command.Command
import org.bukkit.command.CommandExecutor
import org.bukkit.command.CommandSender
import org.bukkit.command.TabCompleter
import org.bukkit.entity.Player

class NodeSessionProfessionRemoveCommand : CommandExecutor, TabCompleter {
    override fun onCommand(
        sender: CommandSender,
        command: Command,
        label: String,
        args: Array<String>,
    ): Boolean {
        if (!sender.hasPermission("numinoustreasury.command.node.create")) {
            sender.sendMessage(ChatColor.RED.toString() + "You do not have permission to create nodes.")
            return true
        }
        if (sender !is Player) {
            sender.sendMessage(ChatColor.RED.toString() + "You must be a player to create nodes.")
            return true
        }
        val nodeService =
            Services.INSTANCE.get(
                NuminousNodeService::class.java,
            )
        var session = nodeService.getNodeCreationSession(sender)
        if (session == null) {
            session = NuminousNodeCreationSession()
            nodeService.setNodeCreationSession(sender, session)
        }
        val professionService =
            Services.INSTANCE.get(
                NuminousProfessionService::class.java,
            )
        if (args.size > 0) {
            val professionName = args.joinToString(" ")
            var profession = professionService.getProfessionById(professionName)
            if (profession == null) {
                profession = professionService.getProfessionByName(professionName)
            }
            if (profession == null) {
                sender.sendMessage(ChatColor.RED.toString() + "There is no profession by that name.")
                return true
            }
            session.removeRequiredProfessionLevel(profession)
            session.display(sender)
        } else {
            sender.sendMessage(ChatColor.WHITE.toString() + "Select a profession: ")
            professionService.professions.forEach { profession ->
                sender.spigot().sendMessage(
                    *ComponentBuilder(profession.name)
                        .color(ChatColor.GRAY)
                        .event(
                            ClickEvent(
                                RUN_COMMAND,
                                "/node session profession remove " + profession.id,
                            ),
                        )
                        .event(
                            HoverEvent(
                                SHOW_TEXT,
                                Text(
                                    ComponentBuilder("Click here to remove " + profession.name)
                                        .color(ChatColor.GRAY)
                                        .create(),
                                ),
                            ),
                        )
                        .create(),
                )
            }
        }
        return true
    }

    override fun onTabComplete(
        sender: CommandSender,
        command: Command,
        label: String,
        args: Array<String>,
    ): List<String> {
        val professionService =
            Services.INSTANCE.get(
                NuminousProfessionService::class.java,
            )
        return if (args.isEmpty()) {
            professionService.professions
                .flatMap { profession ->
                    listOf(
                        profession.id,
                        profession.name,
                    )
                }
        } else {
            professionService.professions
                .flatMap { profession ->
                    listOf(
                        profession.id,
                        profession.name,
                    )
                }
                .filter { name: String ->
                    name.lowercase()
                        .startsWith(args.joinToString(" ").lowercase())
                }
        }
    }
}
